File: interview_0402/cpp/interview_0402.cpp
//
// Created by wangzy on 2022/4/3.
//
#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        /**
        二分来 分别求出有序 数组段 中最短的 子树并返回根节点
        */
        if(nums.size() == 0)
            return nullptr;
        TreeNode* root = rebuildBST(nums,0,nums.size()-1);
        return root;
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    TreeNode* rebuildBST(vector<int>& nums,int start,int end){
        if(start > end)
            return nullptr;
        if(start == end)
            return new TreeNode(nums[start]);
        int mid = start + (end - start)/2;
        int rootVal = nums[mid];
        TreeNode* root = new TreeNode(rootVal);
        root->left = rebuildBST(nums,start,mid-1);
        root->right = rebuildBST(nums,mid+1,end);
        return root;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
//    TreeNode* root = new TreeNode(5);
//    TreeNode* left = new TreeNode(3);
//    TreeNode* right = new TreeNode(6);
//    TreeNode* ll = new TreeNode(2);
//    TreeNode* lr = new TreeNode(4);
//    TreeNode* rl = nullptr;
//    TreeNode* rr = new TreeNode(7);
//    root->left = left;
//    root->right = right;
//    left->left = ll;
//    left->right = lr;
//    right->left = rl;
//    right->right = rr;
    vector<int> nums = {-10,-3,0,5,9};
    //给定有序数组: [-10,-3,0,5,9]
    TreeNode* root = Solution().sortedArrayToBST(nums);
    vector<string> res = Solution().levelOrderStr(root);
    printVec(res);
    //[0,-10,5,null,-3,null,9]
    //一个可能的答案是：[0,-3,9,-10,null,5]，它可以表示下面这个高度平衡二叉搜索树
}

File: interview_0402/java/interview_0402.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class interview_0402 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public TreeNode sortedArrayToBST(int[] nums) {
            /**
             二分来 分别求出有序 数组段 中最短的 子树并返回根节点
             */
            if(nums.length == 0)
                return null;
            return rebuildBST(nums,0,nums.length-1);
        }
        private TreeNode rebuildBST(int[] nums,int start,int end){
            if(start > end)
                return null;
            if(start == end)
                return new TreeNode(nums[start]);
            int mid = start + (end - start)/2;
            TreeNode root = new TreeNode(nums[mid]);
            root.left = rebuildBST(nums,start,mid-1);
            root.right = rebuildBST(nums,mid+1,end);
            return root;
        }
        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        int[] arr = {-10,-3,0,5,9};
        //给定有序数组: [-10,-3,0,5,9],
        Solution solution = new interview_0402().new Solution();
        TreeNode root = solution.sortedArrayToBST(arr);
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        /**
         * 一个可能的答案是：[0,-3,9,-10,null,5]，它可以表示下面这个高度平衡二叉搜索树：
         *
         *           0
         *          / \
         *        -3   9
         *        /   /
         *      -10  5
         *
         *
         *
         */
        //result
        //[0, -10, 5, null, -3, null, 9]
    }
}
